presto-main/src/main/java/com/facebook/presto/operator/Driver.java

```diff
@@ -317,7 +317,7 @@ private void processNewSources()
                 }
 
                 // if current operator is finished...
-                if (current.isFinished()) {
+                if (current.isFinished() && !next.isFinished() && isBlocked(next).isDone()) {
```

Comment:
The only guaranteed method ordering is that before `addInput` is called, `needsInput` will have returned true.  Any operator that depends on `isBlocked` has a bug.  `PartitionedOutputOperator` correctly implements this behavior as follows:
```java
    @Override
    public boolean needsInput()
    {
        return !finished && isBlocked().isDone();
    }

    @Override
    public void addInput(Page page)
    {
        requireNonNull(page, "page is null");
        checkState(isBlocked().isDone(), "output is already blocked");

        if (page.getPositionCount() == 0) {
            return;
        }

        page = pagePreprocessor.apply(page);
        blocked = partitionFunction.partitionPage(page);

        operatorContext.recordGeneratedOutput(page.getSizeInBytes(), page.getPositionCount());
    }
```